### ISSUE TITLE
subs do not autovivify

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#4619]
Migrated from [rt.perl.org#7946](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=7946) (status was 'resolved')

Searchable as RT7946$

**Comment by p5pRT** [comment]
### From joe@sunstarsys.com

 
&nbsp;   
\[originally posted to clpmisc\- RGS suggested I repost it here\]

Why don't the return values of perl subs autovivify?

&nbsp;  % perl \-wle ' sub f \{$\_\} print f\(\) \-> \[0\] '
&nbsp;  Can't use an undefined value as an ARRAY reference at \-e line 1\.

It doesn't even work for lvalued subs&#8203;:

&nbsp;  % perl \-wle ' sub f&#8203;:lvalue \{$\_\} print f\(\) \-> \[0\] '
&nbsp;  Can't use an undefined value as an ARRAY reference at \-e line 1\.

Can this feature be exploited in perl6 to control autovivication?

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=low

Site configuration information for perl v5.6.1:

Configured by joe at Tue Jun 12 14:52:01 EDT 2001.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.3, archname=i586-linux-thread-multi
    uname='linux mumonkan.sunstarsys.com 2.4.3 #3 thu apr 26 11:39:34 edt 2001 i586 unknown '
    config_args=''
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=define use5005threads=undef useithreads=define usemultiplicity=define
    useperlio=define d_sfio=define uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-D_REENTRANT -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/gnu/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2 -g',
    cppflags='-D_REENTRANT -DDEBUGGING -fno-strict-aliasing -I/usr/local/include -I/usr/gnu/include'
    ccversion='', gccversion='2.96 20000731 (Red Hat Linux 7.0)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=y, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib -L/usr/gnu/lib'
    libpth=/usr/local/lib /usr/gnu/lib /lib /usr/lib
    libs=-lsfio -lnsl -lndbm -lgdbm -ldl -lm -lpthread -lc -lposix -lcrypt -lutil
    perllibs=-lsfio -lnsl -ldl -lm -lpthread -lc -lposix -lcrypt -lutil
    libc=/lib/libc-2.1.92.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fpic', lddlflags='-shared -L/usr/local/lib -L/usr/gnu/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /usr/lib/perl5/5.6.1/i586-linux-thread-multi
    /usr/lib/perl5/5.6.1
    /usr/lib/perl5/site_perl/5.6.1/i586-linux-thread-multi
    /usr/lib/perl5/site_perl/5.6.1
    /usr/lib/perl5/site_perl
    /usr/local/lib/perl5/vendor/lib/perl5/vendor_perl/5.6.1/i586-linux-thread-multi
    /usr/local/lib/perl5/vendor/lib/perl5/vendor_perl/5.6.1
    /usr/local/lib/perl5/vendor/lib/perl5/vendor_perl
    .


Environment for perl v5.6.1:
    HOME=/home/joe
    LANG=en_US
    LANGUAGE (unset)
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/usr/kerberos/sbin:/home/joe/bin:/home/joe/bin:/usr/kerberos/bin:/usr/bin:/bin:/usr/X11R6/bin:/usr/local/bin:/opt/bin:/usr/gnu/bin:/usr/X11R6/bin:/usr/bin/mh:/usr/gnu/bin:/usr/X11R6/bin:/usr/bin/mh:/usr/bin/mh:/usr/bin/mh:/usr/bin/mh
    PERL_BADLANG (unset)
    SHELL=/bin/bash

```
</details>

**Comment by p5pRT** [comment]
### From @cpansprout

On Tue Nov 27 03&#8203;:58&#8203;:38 2001\, joe@&#8203;sunstarsys\.com wrote&#8203;:
> Why don't the return values of perl subs autovivify?
> 
>   % perl \-wle ' sub f \{$\_\} print f\(\) \-> \[0\] '
>   Can't use an undefined value as an ARRAY reference at \-e line 1\.

Because normal subs copy their return values\. We can’t change that
without breaking things\.

> It doesn't even work for lvalued subs&#8203;:
> 
>   % perl \-wle ' sub f&#8203;:lvalue \{$\_\} print f\(\) \-> \[0\] '
>   Can't use an undefined value as an ARRAY reference at \-e line 1\.
> 
> Can this feature be exploited in perl6 to control autovivication?

How about 5\.16? That has a six in it\. :\-\)

I’ve just fixed it for lvalue subs with commit 767eda4\.

**Comment by p5pRT** [comment]
@cpansprout - Status changed from 'open' to 'resolved'
